### PR TITLE
Try to use window.edit.Edit when an editable window supports it

### LINKED ISSUE #636: NVDA doesn't speak with NOTETAB text editor
**Reported by marco777 on 2010-04-30 13:15**
Notetab, www.notetab.com , is a good text editor used for many blind people. Infortunatelly NVDA doesn't speak with it. Is possible to do NVDA works with Notetab?

### LINKED ISSUE #8102: Use NVDAObjects.window.edit.Edit overlay class for editable windows with window text
### Description of issue
NVDA uses NVDAObjects.window.edit.Edit for windows with the "edit" window class. The associated TextInfo class is EditTextInfo for non rich edit cases. This NVDAObject class relies on the NVDAObjects.window.Window.windowText to get the text content. For plain text content, this NVDAObject class functions quite well.

For several text windows, NVDA now relies on the display model to get text contents. Examples of those are text controls with specific window classes in delphi applications. Though displayModel.EditableTextDisplayModelTextInfo has the advantage of providing bold, italic and color information, it also has the down side of being unreliable in some cases, especially when inserting or deleting text. It is important to note that, in several delphi text controls I stumbled upon in the wild also expose the text using the windowText property on the object. It is also possible to get selection offsets with the EM_GETSEL window message.

## Proposed fix
In Window.findOverlayClasses, before falling back to the DisplayModelEditableText, send the EM_GETLINECOUNT message to the window. If greater than 0, use a new UnidentifiedEdit NVDAObject class instead of the DisplayModelEditableText. The API version for this UnidentifiedEdit class could be set to -1, and the use of iTextDocument should be attempted when constructing a TextInfo for this class. Note the following information about EM_GETLINECOUNT from [msdn](https://msdn.microsoft.com/en-us/library/windows/desktop/bb761586(v=vs.85).aspx):

> The return value is an integer specifying the total number of text lines in the multiline edit control or rich edit control. If the control has no text, the return value is 1. The return value will never be less than 1.

Testing reveals that this window message also returns 1 for single line edit controls.

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Closes #8102
Closes #636

### Summary of the issue:
NVDA uses the display model for editable text cases where there is a caret in the window and there doesn't seem to be another overlay class that implements editing. However, NVDA is a bit conservative in this, only using window.edit.Edit for objects with a particular window class or for windowless richEdit.

### Description of how this pull request fixes the issue:
When finding an overlay class for an object and there is a caret in the window, rather than directly choosing DisplayModelEditableText, this pr will send the EM_LINELENGTH message to the window. This window message always returns 1 or higher for windows that support the edit API, also for single line edit controls. This means that, when the window responds with >=1 to EM_LINELENGTH, the edit API is supported. In this case, a new UnidentifiedEdit class is used that makes no assumptions about the edit API version. For these controls, this pr will try to construct an ITextDocumentTextInfo object to retrieve text information. If that fails, the implementation will fall back to EditTextInfo.

### Testing performed:
Tested the notetab text editor. Instead of using DisplayModelEditableText for the edit control, UnidentifiedEdit is now used, and the control even implements ITextDocument support. Also tested a corporate Delphi application that has edit controls with custom window classes.

### Known issues with pull request:
None

### Change log entry:
* Bug fixes
    - In some text controls, particularly in Delphi applications, the information provided about editing and navigating is now much more reliable. (#636, #8102)

### CODE REVIEW

File: source/NVDAObjects/IAccessible/__init__.py [hunk 1]
```diff
@@ -511,7 +511,6 @@ def findOverlayClasses(self,clsList):
 				pDoc=None
 			if pDoc:
 				self._ITextDocumentObject=pDoc
-				self.useITextDocumentSupport=True
```
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/__init__.py]: Why was this removed?

File: source/NVDAObjects/window/edit.py [hunk 2]
```diff
@@ -379,38 +368,6 @@ def _getWordOffsets(self,offset):
 				start=end
 				end=watchdog.cancellableSendMessage(self.obj.windowHandle,EM_FINDWORDBREAK,WB_MOVEWORDRIGHT,offset)
 			return (start,end)
-		elif winVersion.winVersion.major<6: #Implementation of standard edit field wordbreak behaviour (only breaks on space)
```
Comment (michaelDCurran) [inline source/NVDAObjects/window/edit.py]: Why is all this code removed?

**Review by michaelDCurran — COMMENTED** [review]
This pr is rather hard to follow due to the large amount of constants moved to winUser. I understand why you did it, but  there are various bits of code here which I'm confused as to why they are removed.

File: source/NVDAObjects/IAccessible/__init__.py (hunk 1 above)
Comment (leonardder) [reply source/NVDAObjects/IAccessible/__init__.py]: I will explain this below.

File: source/NVDAObjects/window/edit.py (hunk 2 above)
Comment (leonardder) [reply source/NVDAObjects/window/edit.py]: Because it is code that would only apply to Windows XP.

File: source/NVDAObjects/window/edit.py [hunk 3]
```diff
@@ -773,13 +730,10 @@ def updateSelection(self):
 class Edit(EditableTextWithAutoSelectDetection, Window):
 
 	editAPIVersion=0
-	editAPIUnicode=True
-	# #4291: Use ITextDocument in Windows 7 and later, as it's very slow in earlier versions.
-	useITextDocumentSupport=(winVersion.winVersion.major,winVersion.winVersion.minor)>=(6,1)
```
Comment (leonardder) [inline source/NVDAObjects/window/edit.py]: @MichaelDCurran: the useITextDocumentSupport property seems to have been there to limit the support of ITextDocument to windows 7 and abbove. Since we aren't longer supporting Windows versions older than 7, this code now became obsolete (i.e. this variable would be always True). That's also the reason why I removed it from the IAccessible windowless rich text edit support as you saw.

**Review by michaelDCurran — APPROVED** [review]

**Comment by Qchristensen** [comment]
One issue I've found from this PR, is moving the mouse while Notepad is open causes a lovely staccato of the error sound.  The error it's generating is:

ERROR - eventHandler.executeEvent (13:00:27.961):
error executing event: mouseMove on <NVDAObjects.Dynamic_IAccessibleEditWindowNVDAObject object at 0x04F52B50> with extra args of {'y': 314, 'x': 953}
Traceback (most recent call last):
  File "eventHandler.pyc", line 152, in executeEvent
  File "eventHandler.pyc", line 92, in __init__
  File "eventHandler.pyc", line 100, in next
  File "NVDAObjects\__init__.pyc", line 923, in event_mouseMove
  File "documentBase.pyc", line 24, in makeTextInfo
  File "textInfos\offsets.pyc", line 282, in __init__
  File "NVDAObjects\window\edit.pyc", line 196, in _getOffsetFromPoint
NameError: global name 'EM_CHARFROMPOS' is not defined

**Comment by leonardder** [comment]
[**@Qchristensen**](https://github.com/Qchristensen) commented on [17 Apr 2018, 05:25 CEST](https://github.com/nvaccess/nvda/pull/8165#issuecomment-381823464 "2018-04-17T03:25:17Z - Replied by Github Reply Comments"):
> One issue I've found from this PR, is moving the mouse while Notepad is open causes a lovely staccato of the error sound.

Fixed in https://github.com/nvaccess/nvda/commit/ebfe710 .

**Comment by Qchristensen** [comment]
That's behaving better, thanks Leonard!

**Comment by beqabeqa473** [comment]
hello.

maybe after this pull request i get this
ImportError: cannot import name EM_SETSEL


i wrote spell checker addon which gets and sets selection.
after some update i am getting this error and addon no longer works.
is it by design or it is a miss?

thanks.

**Comment by michaelDCurran** [comment]
Those constants moved from NVDAObjects.window.edit to winUser.

**Comment by leonardder** [comment]
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [3 mei 2018 09:04 CEST](https://github.com/nvaccess/nvda/pull/8165#issuecomment-386207482 "2018-05-03T07:04:57Z - Replied by Github Reply Comments"):
> Those constants moved from NVDAObjects.window.edit to winUser.

I guess we could revert this change, but as Window.windowTextLineCount not relies on one of these constants, the edit module didn't seem to be the right place for it. Add to that the fact that these constants live in winUser.h in the windows SDK.
